packages/flutter/test/widgets/two_dimensional_viewport_test.dart

```diff
@@ -0,0 +1,182 @@
+    return RenderSimpleTableViewport(
+      horizontalOffset: horizontalOffset,
+      horizontalAxisDirection: horizontalAxisDirection,
+      verticalOffset: verticalOffset,
+      verticalAxisDirection: verticalAxisDirection,
+      mainAxis: mainAxis,
+      delegate: delegate,
+      childManager: context as RenderObjectElement,
+      cacheExtent: cacheExtent,
+      clipBehavior: clipBehavior,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, RenderSimpleTableViewport renderObject) {
+    renderObject
+      ..horizontalOffset = horizontalOffset
+      ..horizontalAxisDirection = horizontalAxisDirection
+      ..verticalOffset = verticalOffset
+      ..verticalAxisDirection = verticalAxisDirection
+      ..mainAxis = mainAxis
+      ..delegate = delegate
+      ..cacheExtent = cacheExtent
+      ..clipBehavior = clipBehavior;
+  }
+}
+
+class RenderSimpleTableViewport extends RenderTwoDimensionalViewport {
+  RenderSimpleTableViewport({
+    required super.horizontalOffset,
+    required super.horizontalAxisDirection,
+    required super.verticalOffset,
+    required super.verticalAxisDirection,
+    required super.delegate,
+    required super.mainAxis,
+    required super.childManager,
+    super.cacheExtent,
+    super.clipBehavior = Clip.hardEdge,
+    this.applyDimensions = true,
+    this.useParentSize = true,
+    this.setLayoutOffset = true,
+  });
+
+  final bool applyDimensions; // Testing error message
+  final bool useParentSize; // Testing error message
+  final bool setLayoutOffset; // Testing error message
+
+  @override
+  void layoutChildSequence() {
+    // Really simple table implementation for testing.
+    // Every child is 200x200 square
+    final double horizontalPixels = horizontalOffset.pixels;
+    final double verticalPixels = verticalOffset.pixels;
+    final int leadingColumn = math.max((horizontalPixels / 200).floor(), 0);
+    final int leadingRow = math.max((verticalPixels / 200).floor(), 0);
+    final int trailingColumn = math.min(((horizontalPixels + viewportDimension.width) / 200).ceil(), 99);
+    final int trailingRow = math.min(((verticalPixels + viewportDimension.height) / 200).ceil(), 99);
+
+    double xLayoutOffset = (leadingColumn * 200) - horizontalOffset.pixels;
+    for(int column = leadingColumn; column <= trailingColumn; column++) {
+      double yLayoutOffset =  (leadingRow * 200) - verticalOffset.pixels;
+      for (int row = leadingRow; row <= trailingRow; row++) {
+        final RenderBox child = buildOrObtainChildFor(ChildVicinity(xIndex: row, yIndex: column))!;
+        child.layout(constraints, parentUsesSize: useParentSize);
+
+        if (setLayoutOffset) {
+          parentDataOf(child).layoutOffset = Offset(xLayoutOffset, yLayoutOffset);
+        }
+        yLayoutOffset += 200;
+      }
+      xLayoutOffset += 200;
+    }
+    if (applyDimensions) {
+      verticalOffset.applyContentDimensions(0, 200 * 100 - viewportDimension.height);
```

Comment:
This is the only thing I am still unsure of. I tried to write some logic to estimate the max scroll extent for the user, but it was really hard to guess what they were building or how long it would be in one or both dimensions.

So right now, the user has to call applyContentDimensions. I currently have a check that will tell them to do so if they have not.

The other thing I considered was exposing a `maxScrollOffset` getter that the subclasses would have to implement, and then the super class could call applyContentDimensions with that at the end of performLayout.

I like the getter approach, curious to know what you think first though. :)

Reply:
Either seems fine to me. If we want people to call applyContentDimensions within layoutChildSequence we should document that, though. The getter may be a little less flexible (which could be fine), but simpler to implement.